### PR TITLE
macOS installer

### PR DESCRIPTION
**Pull request by asimilon**

### CODE REVIEW

**Comment by baconpaul** [comment]
Thank you for doing this work. I will take a look in the next day or so. Quick scan from my phone looks promising but I’d like to pull and try it before I merge.

**Comment by asimilon** [comment]
Yes.  I tested as well I could, but would be nice for more people to confirm it's doing everything OK.

**Comment by asimilon** [comment]
Perhaps this is more readable :
```
# create distribution.xml

if [[ -d $VST2 ]]; then
	VST2_PKG_REF='<pkg-ref id="com.vemberaudio.vst2.pkg"/>'
	VST2_CHOICE='<line choice="com.vemberaudio.vst2.pkg"/>'
	VST2_CHOICE_DEF="<choice id=\"com.vemberaudio.vst2.pkg\" visible=\"true\" title=\"Install VST2\"><pkg-ref id=\"com.vemberaudio.vst2.pkg\"/></choice><pkg-ref id=\"com.vemberaudio.vst2.pkg\" version=\"${VERSION}\" onConclusion=\"none\">Surge_VST2.pkg</pkg-ref>"
fi
if [[ -d $VST3 ]]; then
	VST3_PKG_REF='<pkg-ref id="com.vemberaudio.vst3.pkg"/>'
	VST3_CHOICE='<line choice="com.vemberaudio.vst3.pkg"/>'
	VST3_CHOICE_DEF="<choice id=\"com.vemberaudio.vst3.pkg\" visible=\"true\" title=\"Install VST3\"><pkg-ref id=\"com.vemberaudio.vst3.pkg\"/></choice><pkg-ref id=\"com.vemberaudio.vst3.pkg\" version=\"${VERSION}\" onConclusion=\"none\">Surge_VST3.pkg</pkg-ref>"
fi
if [[ -d $AU ]]; then
	AU_PKG_REF='<pkg-ref id="com.vemberaudio.au.pkg"/>'
	AU_CHOICE='<line choice="com.vemberaudio.au.pkg"/>'
	AU_CHOICE_DEF="<choice id=\"com.vemberaudio.au.pkg\" visible=\"true\" title=\"Install Audio Unit\"><pkg-ref id=\"com.vemberaudio.au.pkg\"/></choice><pkg-ref id=\"com.vemberaudio.au.pkg\" version=\"${VERSION}\" onConclusion=\"none\">Surge_AU.pkg</pkg-ref>"
fi

cat > distribution.xml << XMLEND
<?xml version="1.0" encoding="utf-8"?>
<installer-gui-script minSpecVersion="1">
    <title>Install Surge ${VERSION}</title>
    <license file="License.txt" />
    ${VST2_PKG_REF}
    ${VST3_PKG_REF}
    ${AU_PKG_REF}
    <pkg-ref id="com.vemberaudio.resources.pkg"/>
    <options require-scripts="false"/>
    <choices-outline>
        ${VST2_CHOICE}
        ${VST3_CHOICE}
        ${AU_CHOICE}
        <line choice="com.vemberaudio.resources.pkg"/>
    </choices-outline>
    ${VST2_CHOICE_DEF}
    ${VST3_CHOICE_DEF}
    ${AU_CHOICE_DEF}
    <choice id="com.vemberaudio.resources.pkg" visible="false" title="Install resources">
        <pkg-ref id="com.vemberaudio.resources.pkg"/>
    </choice>
    <pkg-ref id="com.vemberaudio.resources.pkg" version="${VERSION}" onConclusion="none">Surge_Resources.pkg</pkg-ref>
</installer-gui-script>
XMLEND
```

**Comment by baconpaul** [comment]
Yeah that’s a bit easier to read.

I was also thinking - and up to you - perhaps we should name the DMG something like

```
DATE=`date "+%Y-%m-%d"`
DMG_NAME=Surge-$DATE.dmg
```

So we get Surge-2018-12-18.dmg or what not. If you are going to do an update anyway, perhaps consider something like that? You could even do date and time if you wanted. 

And maybe also, if you are at it, just echo date time osversion and build host into resources in a file called “buildinfo.txt” or some such so if someone has a question about a random install we can sort of find out what build they had?

Basically from reviewing your code: I love it and it is super! But if it gets run a lot how do we want to distinguish between resulting assets. (I have not had a chance to run it yet, just read it; I will do that soon enough)

File: installer_osx/make_installer.sh [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+# build resources package
+
+pkgbuild --analyze --root "$RSRCS" Surge_Resources.plist
+pkgbuild --root "$RSRCS" --component-plist Surge_Resources.plist --identifier "com.vemberaudio.resources.pkg" --version $VERSION --scripts ResourcesPackageScript --install-location "/tmp/Surge" Surge_Resources.pkg
+rm Surge_Resources.plist
+
+# create distribution.xml
+
+echo '<?xml version="1.0" encoding="utf-8"?>' > distribution.xml
```
Comment (kzantow) [inline installer_osx/make_installer.sh]: just a suggestion here, maybe use a heredoc to make this more readable? e.g.:

```xml
cat << EOF >> distribution.xml
<?xml version="1.0" encoding="utf-8"?>
<installer-gui-script minSpecVersion="1">
  <title>Install Surge $VERSION</title>
  <license file="License.txt" />
  <pkg-ref id="$PKGREF"/>
...
```

And just use the `if` blocks to set up the necessary variables

**Comment by asimilon** [comment]
I've got waiting to commit also something similar to that suggestion for the `.dmg` name
```
hdiutil convert /tmp/tmp.dmg -format UDZO -o "Install_Surge_$VERSION.dmg"
```
so end up with this result :
image

File: installer_osx/make_installer.sh (hunk 1 above)
Comment (asimilon) [reply installer_osx/make_installer.sh]: hehe, yeah I already thought that after looking at the mess I created :)  there's a comment on the PR itself

**Comment by kzantow** [comment]
@asimilon this is looking excellent! (and now I read your comment, same thing about the heredoc. I would definitely do that, and call this one good after (with testing etc. of course) :)

**Comment by asimilon** [comment]
great, I'll just add the suggestion from @baconpaul about the `buildinfo.txt` in the resource folder and commit those changes

**Comment by asimilon** [comment]
Something like this maybe?
```
# write build info to resources folder

echo "Version: ${VERSION}" > "$RSRCS/BuildInfo.txt"
echo "Built on: $(date "+%Y-%m-%d %H:%M:%S")" >> "$RSRCS/BuildInfo.txt"
echo "On host: $(hostname)" >> "$RSRCS/BuildInfo.txt"
```
anything else required?

**Comment by kzantow** [comment]
@asimilon current SHA would be excellent, too

**Comment by asimilon** [comment]
Could you give an example please?

I'm cleaning up that `BuildInfo.txt` afterwards too, so not to leave pollution :)

**Comment by kzantow** [comment]
Regarding `hdiutil convert /tmp/tmp.dmg -format UDZO -o "Install_Surge_$VERSION.dmg"` -- is there a reason to package something into a `.dmg` when there already is a single `.pkg` file?

**Comment by asimilon** [comment]
So, just drop the `.dmg` entirely?

and add
```
echo "Commit: $(git rev-parse HEAD)" >> "$RSRCS/BuildInfo.txt"
```

**Comment by asimilon** [comment]
Zipping the `.pkg` shaves off 1.5MB, but I think that can be part of deploy step if necessary.

**Comment by baconpaul** [comment]
Yes on that echo.

On the dmg, leave it in since you have written it. I think much of the software I download is a dmg of a pkg but I could be imagining that. But I don’t see harm in leaving in the option and if we don’t use it we can take it out later.

Thank you

**Comment by esaruoho** [comment]
how do i test this? i'm not savvy with "moving to an outside-kurasu-surge-fork branch"..

**Comment by baconpaul** [comment]
@esaruoho when I pull this into my user area later tonight I will keep the log of exact git commands so you can test it also. But basically you add surge as upstream and then git fetch upstream then git checkout upstream/<fork-name> and you are good.

But also I think we will merge this soon. It works for @asimilon and @kzantow and once the final diffs are in I’ll confirm it works for me. If it does I will press merge and then you just do a normal update.

**Comment by asimilon** [comment]
This should do the trick
```
git remote add asimilon https://github.com/asimilon/surge
git fetch asimilon
git checkout asimilon/osx_installer
```

I just pushed the final changes.

**Comment by asimilon** [comment]
I tend to write a script like this when I'm developing plugins, just to be sure everything is right
```
$ cat trash_surge.sh 
#!/bin/bash
sudo rm -rf /Library/Audio/Plug-Ins/Components/Surge.component
sudo rm -rf /Library/Audio/Plug-Ins/VST/Surge.vst
sudo rm -rf /Library/Audio/Plug-Ins/VST3/Surge.vst3
sudo rm -rf ~/Library/Audio/Plug-Ins/Components/Surge.component
sudo rm -rf ~/Library/Audio/Plug-Ins/VST/Surge.vst
sudo rm -rf ~/Library/Audio/Plug-Ins/VST3/Surge.vst3
sudo rm -rf ~/Library/Application\ Support/Surge
```

**Comment by esaruoho** [comment]
that's a good script - could be added to the repo, right?

**Comment by esaruoho** [comment]
@baconpaul how would this be run by a non-savvy github-newbie like me?
./makefile?
make makefile?

anyway, i'm all for any simplification. still can't get vst2 baked on my own though, which irks me some.

**Comment by baconpaul** [comment]
those would be exactly the commands, if we wrote Makefile.mac.

From the other PR, it looks like you are sometimes running as root. You aren't doing these scripts as sudo ever are you?

**Comment by esaruoho** [comment]
@baconpaul i never really mess around the terminal as root.. but yeah, i can move stuff to /Library and get prompted for a password.

**Comment by baconpaul** [comment]
Yeah I was thinking the same.

I was also thinking of making build-osx.sh a bit of a meatier script with functions and options. So you could do

```
./build-osx.sh  # current behaviour
./build-osx.sh --install-local-au  # basically what my install local au does
--install-local-vst
--install-local
--clean-builds
--remove-installs
--build-pkg
```

you get the idea. Then we can just add functions to that on switches. Easy enough.

**Comment by baconpaul** [comment]
I made a fresh issue for this convo rather than an old PR.

**Comment by asimilon** [comment]
I like that idea for the build flags 👍
